components/layout_2020/display_list/clip_path.rs

```diff
@@ -0,0 +1,228 @@
+                top_right: radius,
+                bottom_left: radius,
+                bottom_right: radius,
+            };
+            let start = center.add_size(&-radius);
+            let rect = LayoutRect::from_origin_and_size(start, radius * 2.);
+            normalize_radii(&layout, &mut radii);
+            create_rect_clip_chain(
+                radii,
+                rect,
+                parent_scroll_node_id,
+                parent_clip_chain_id,
+                display_list,
+            )
+        },
+        BasicShape::Ellipse(ellipse) => {
+            let center = match ellipse.position {
+                GenericPositionOrAuto::Position(position) => position,
+                GenericPositionOrAuto::Auto => Position::center(),
+            };
+            let anchor_x = center.horizontal.resolve(Length::new(layout.width()));
+            let anchor_y = center.vertical.resolve(Length::new(layout.height()));
+            let amount = LayoutSize::new(anchor_x.px(), anchor_y.px());
+            let center = layout.min.add_size(&amount);
+            let width = if let GenericShapeRadius::Length(length) = ellipse.semiaxis_x {
+                length.0.resolve(Length::new(layout.width())).px()
+            } else {
+                compute_shape_radius(center.x, &ellipse.semiaxis_x, layout.min.x, layout.max.x)
+            };
+            let height = if let GenericShapeRadius::Length(length) = ellipse.semiaxis_y {
+                length.0.resolve(Length::new(layout.height())).px()
+            } else {
+                compute_shape_radius(center.y, &ellipse.semiaxis_y, layout.min.y, layout.max.y)
+            };
+            let mut radii = webrender_api::BorderRadius {
+                top_left: LayoutSize::new(width, height),
+                top_right: LayoutSize::new(width, height),
+                bottom_left: LayoutSize::new(width, height),
+                bottom_right: LayoutSize::new(width, height),
+            };
+            let size = LayoutSize::new(width, height);
+            let start = center.add_size(&-size);
+            let rect = LayoutRect::from_origin_and_size(start, size * 2.);
+            normalize_radii(&rect, &mut radii);
+            create_rect_clip_chain(
+                radii,
+                rect,
+                parent_scroll_node_id,
+                parent_clip_chain_id,
+                display_list,
+            )
+        },
+        _ => None,
+    }
+}
+
+fn compute_shape_radius(
+    center: f32,
+    radius: &GenericShapeRadius<NonNegativeLengthPercentage>,
+    layout_min: f32,
+    layout_max: f32,
+) -> f32 {
+    let left = (layout_min - center).abs();
+    let right = (layout_max - center).abs();
+    match (radius, left.partial_cmp(&right)) {
+        (GenericShapeRadius::FarthestSide, Some(Ordering::Greater)) => left,
+        (GenericShapeRadius::FarthestSide, _) => right,
+        (_, Some(Ordering::Greater)) => right,
+        (_, _) => left,
+    }
```

Comment:
I think this could be simplified:
```suggestion
    match radius {
        GenericShapeRadius::FarthestSide => left.max(right),
        GenericShapeRadius::ClosestSide => left.min(right),
    }
```